### PR TITLE
Disallow duplicated device name for the same user

### PR DESCRIPTION
**Pull request by kailichiang**
## Description

Disallow and show danger message when
- Add a device using an exiting device name.
- Update a device name to another existing device name.

## Motivation and Context

Disallow duplicated device name for the same user #162

## How Has This Been Tested?

Add a new test case for adding a device with an existing name.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Enhancement/Security (non-breaking change which is not noticeable to end users)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I created a feature branch and did not open a pull request from my `master` branch.
- [x] My code follows the code style of this project.
- [ ] My change requires an update to the README and I have updated it accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] This is a complete change and doesn't leave the project in a bad state.
- [x] All new and existing tests pass.

### CODE REVIEW

File: app/Http/Controllers/Web/DevicesController.php [hunk 1]
```diff
@@ -38,6 +40,22 @@ public function add(Request $request): RedirectResponse
     {
         $properties = $request->all();
         $currentUserId = $request->user()->id;
+
+        $data = array_merge($properties, ['user_id' => $currentUserId]);
+        $validator = Validator::make( $data, [
```
Comment (dbudwin) [inline app/Http/Controllers/Web/DevicesController.php]: Please remove the leading space before `$data`.

File: app/Http/Controllers/Web/DevicesController.php [hunk 2]
```diff
@@ -75,6 +93,23 @@ public function update(Request $request, Uuid $publicDeviceId): RedirectResponse
         }
 
         $properties = $request->all();
+        $currentUserId = $request->user()->id;
+
+        $data = array_merge($properties, ['user_id' => $currentUserId, 'device_id' => $deviceId]);
+        $validator = Validator::make( $data, [
```
Comment (dbudwin) [inline app/Http/Controllers/Web/DevicesController.php]: Please remove the leading space before `$data`.

**Review by dbudwin — CHANGES_REQUESTED** [review]
Looks good!  Mainly just some simple requests.  Also take a look at the results from the linter (don't worry about the lines being too long messages): https://codeclimate.com/github/dbudwin/RoboHome-Web/pull/164

**Comment by dbudwin** [comment]
I pulled this code down and played with it.  It does work to prevent duplicate device names.  However, I think it would be a better user experience if the user was given a chance to correct the device name instead of just seeing an error message.  For example, when a user is creating a device named "Test" and they click "Add Device," instead of the modal closing and the flash message being displayed, they should get a validation error within the modal where they can change the device name to something else without losing the rest of their settings.

**Comment by kailichiang** [comment]
About "better experience," there are several possible solutions.

1. Use ajax to validate input value before really creating or updating. This is gonna be great user experience. But it needs lots of js code, a new API, and so on.
2. After backend validation and redirection, a user sees error message. When clicking the same button, the modal shows with old values. This is gonna be the simplest.
3. After backend validation, a user redirect back with modal opening (if any error) with old values. This needs a logic or mechanism to know why a user back to determine opening modal or not, showing add-modal or edit-modal etc.

Solution 1 and 3 probably needs lots engineering, I'm not sure. You may have more experience.

**Comment by dbudwin** [comment]
I agree, I think option 1 would have the best user experience.  I don't have a ton of JS experience (I try to avoid it as much as I can), but I figured there has to be a JS library out there that could help.  Doing some searching I found [Parsley](http://parsleyjs.org/).  There was also a recent question on [StackOverflow](https://stackoverflow.com/questions/50999599/parsley-2-8-1-unique-username-validation-with-php) trying to solve a problem very similar to this one.  The same user as that StackOverflow question also got some guidance when asked on [Parsley's GitHub repository](https://github.com/guillaumepotier/Parsley.js/issues/1258) too.

In brief, it looks like we would need to use [Parsley's Remote Validator](http://parsleyjs.org/doc/index.html#remote) to hit a new API endpoint that would return a list of all the device names that a user has.  So, if I understand correctly, we shouldn't need much more than a small JS function which uses Parsley and a simple API endpoint.  This project already has an API implemented for querying all the devices that a user has for example ([see `index` of `/API/DevicesController.php`](https://github.com/dbudwin/RoboHome-Web/blob/master/app/Http/Controllers/API/DevicesController.php)).

Thoughts?

**Comment by dbudwin** [comment]
I think below the input makes sense.  That seems to be the common way to use that library and their [examples online](http://parsleyjs.org/doc/examples/customvalidator.html) often do that.  It would also make it easier in the future if I needed more validation on some of those fields.

**Comment by dbudwin** [comment]
Are you still considering getting client-side validation working for this issue?

**Comment by kailichiang** [comment]
I’d like to do client-side validation. As you suggested in another PR, before starting, I’m better to delete my original fork and fork again. This PR probably will be ignored. But I’m kinda busy recently, so I don’t start yet. 

The client-side validation can be also documented to the original issue, therefore anyone interested is welcome to do this.

**Pull request closed without merging**